### PR TITLE
[BEAM-658] Support Read.Unbounded primitive.

### PR DESCRIPTION
**Pull request by amitsela**
Be sure to do all of the following to help us incorporate your contribution
quickly and easily:

 - [x] Make sure the PR title is formatted like:
   `[BEAM-<Jira issue #>] Description of pull request`
 - [x] Make sure tests pass via `mvn clean verify`. (Even better, enable
       Travis-CI on your fork and ensure the whole test matrix passes).
 - [x] Replace `<Jira issue #>` in the title with the actual Jira issue
       number, if there is one.
 - [x] If this contribution is large, please file an Apache
       [Individual Contributor License Agreement](https://www.apache.org/licenses/icla.txt).

---

Changed mapSourceFunction to use scala's native Option.

Upgrade to kryo-serializers 0.39 that provides support for ReverseList (used by Top).

Better logging.

Allow a longer read time-frame for read in tests.

Assert initial parallelism is gerater than zero.

Add OnBatchCompleted listener that writes to Kafka.

Test latest.

### CODE REVIEW

**Comment by amitsela** [comment]
R: @dhalperi / @kennknowles 
@jbonofre PTAL, you've tested this while it was WIP with UnboundedSources other then Kafka.

**Comment by amitsela** [comment]
Seems like it complains it can't find `org.apache.beam.sdk.io.BoundedReadFromUnboundedSource` referenced [here](https://github.com/amitsela/incubator-beam/blob/BEAM-658/runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java#L40).
Anybody got an idea why ?

**Comment by dhalperi** [comment]
Hmm. I checked out htis PR and ran `mvn clean verify -DskipTests` and build succeeded.

Trying again with `-Prelease` also set.

**Comment by dhalperi** [comment]
Was able to replicate with `-Prelease` set.

The problem is that `BoundedReadFromUnboundedSource` is package-private. That's a bug -- there's no reason to believe that only code inside of the `sdk.io` package should want to use this functionality. [Send a PR](https://github.com/apache/incubator-beam/pull/1145)

**Comment by amitsela** [comment]
Thanks! I'll rebase on master once merged.

**Comment by amitsela** [comment]
Rebased following #1145

**Comment by amitsela** [comment]
There's a bunch of failures, for the Spark runner it's for tests who checkpoint the Spark job, which is not a must for all of them, but it is for some of them.
This looks like Jenkins is very very slow, so now computation happens by the time the hard timeout is applied (mostly 1 sec).
I could extend this timeout, this will affect all builds, even when Jenkins is not lagging behind.

**Comment by amitsela** [comment]
Rebased & squashed no that #1161 is merged.

**Comment by amitsela** [comment]
This is still not very deterministic, because there are no triggers involved, so if execution is slow elements might not end-up in the same microbatch, or the job will finish before reading all.
Yet another great reason to keep going forward with triggers & accumulation modes.
For now, I'll giving more time on read from Kafka.

**Comment by amitsela** [comment]
Spark tests pass now, it's `Maven Archetypes :: Starter` that is broken (I think it's resolved in master now), I won't rebase again for now as I expect some comments for this PR.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistrator.java [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+import de.javakaffee.kryoserializers.guava.ImmutableMultimapSerializer;
+import de.javakaffee.kryoserializers.guava.ImmutableSetSerializer;
+import de.javakaffee.kryoserializers.guava.ReverseListSerializer;
+import org.apache.spark.serializer.KryoRegistrator;
+
+
+/**
+ * Custom {@link com.esotericsoftware.kryo.Serializer}s for Beam's Spark runner needs.
+ */
+public class BeamSparkRunnerRegistrator implements KryoRegistrator {
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistrator.java]: In terms of review latency, you could probably break out these serialization things into their own commits or separate PR and get it in right away.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+      backoffFactory =
+          FluentBackoff.DEFAULT
+              .withInitialBackoff(Duration.millis(10))
+              .withMaxBackoff(maxReadTime.minus(1))
+              .withMaxCumulativeBackoff(maxReadTime.minus(1));
+    }
+
+    @Override
+    public boolean start() throws IOException {
+      LOG.info("MicrobatchReader-{}: Starting a microbatch read from an unbounded source with a "
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java]: Seems like `LOG.debug` might be more appropriate for this routine and frequent operation.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java [hunk 3]
```diff
@@ -0,0 +1,262 @@
...
+        Uninterruptibles.sleepUninterruptibly(nextSleep, TimeUnit.MILLISECONDS);
+        nextSleep = backoff.nextBackOffMillis();
+      }
+      finalizeCheckpoint();
+      return false;
+    }
+
+    private void finalizeCheckpoint() throws IOException {
+      reader.getCheckpointMark().finalizeCheckpoint();
+      LOG.info("MicrobatchReader-{}: finalized CheckpointMark successfully after "
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/io/MicrobatchSource.java]: Perhaps `LOG.debug` here as well.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceDStream.java [hunk 4]
```diff
@@ -0,0 +1,156 @@
...
+import org.joda.time.Duration;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+
+import scala.Tuple2;
+
+
+
+/**
+ * A SourceDStream is an {@link InputDStream} of {@link SourceRDD.Unbounded}s.
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceDStream.java]: Nit: `{@link SourceDStream}`, etc throughout for linking but also IDE/compiler friendliness.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceDStream.java [hunk 5]
```diff
@@ -0,0 +1,156 @@
...
+
+  @Override
+  public void start() { }
+
+  @Override
+  public void stop() { }
+
+  @Override
+  public String name() {
+    return "Beam UnboundedSource [" + id() + "]";
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceDStream.java]: Recommend `MoreObjects.toStringHelper` as a nice standardized format.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java [hunk 6]
```diff
@@ -196,5 +197,75 @@ public int index() {
...
       return 41 * (41 + rddId) + index;
     }
+
+    public Source<T> getSource() {
+      return source;
+    }
+  }
+
+  /**
+   * A SourceRDD.Unbounded is the implementation of a micro-batch in a {@link SourceDStream}.
```
Comment (kennknowles) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/io/SourceRDD.java]: Nit: `{@link SourceRDD.Unbounded}`

File: runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/KafkaStreamingTest.java [hunk 7]
```diff
@@ -53,32 +56,14 @@
  */
 public class KafkaStreamingTest {
   private static final EmbeddedKafkaCluster.EmbeddedZookeeper EMBEDDED_ZOOKEEPER =
-          new EmbeddedKafkaCluster.EmbeddedZookeeper();
+      new EmbeddedKafkaCluster.EmbeddedZookeeper();
```
Comment (kennknowles) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/KafkaStreamingTest.java]: Now that you support `Read.Unbounded` can you use `KafkaIO` ROS tests? I know they have an unclear interpretation (they only work with embedded Kafka & embedded runner, last time I was involved) but seems to apply to this case.

Comment (kennknowles) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/KafkaStreamingTest.java]: Separately, for just testing streaming aspects of the runner, perhaps `CountingSource` is a more agnostic choice?

File: runners/spark/pom.xml [hunk 8]
```diff
@@ -112,29 +112,11 @@
     </dependency>
     <dependency>
       <groupId>org.apache.spark</groupId>
-      <artifactId>spark-streaming-kafka_2.10</artifactId>
```
Comment (kennknowles) [inline runners/spark/pom.xml]: (Commenting here because related lines not part of the diff) Can you add a `-Plocal-runnable-on-service-tests-streaming` profile for easy launching?

Comment (amitsela) [inline runners/spark/pom.xml]: Opened [BEAM-785](https://issues.apache.org/jira/browse/BEAM-785) as a separate issue, to follow after this PR.
Probably most of the tests would be ignored at first (due to lack of support in triggers, accumulations, etc.) but it'd be nice to have something to start with.

File: runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/KafkaStreamingTest.java (hunk 7 above)
Comment (amitsela) [inline runners/spark/src/test/java/org/apache/beam/runners/spark/translation/streaming/KafkaStreamingTest.java]: Once I have confidence that the runner's streaming support is clear enough (meaning a set of some ROS tests running) I'll consider.
I am however all for being "Kafka-aware" as this will probably be the main `UnboundedSource` users will use.

File: runners/spark/src/main/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistrator.java (hunk 1 above)
Comment (amitsela) [inline runners/spark/src/main/java/org/apache/beam/runners/spark/coders/BeamSparkRunnerRegistrator.java]: Ported to #1171.

**Comment by dhalperi** [comment]
R: -@dhalperi @kennknowles 

Handing off to Kenn :)

**Comment by kennknowles** [comment]
OK, this is forward progress. I agree that the Jenkins signal does not apply to this PR.

So LGTM, with misgivings about our level of testing signal :-/

**Comment by amitsela** [comment]
I'll relate to the comments and handle #1171 before merging so Jenkins will have another chance 😉

**Comment by amitsela** [comment]
Rebased on master, once Jenkins passes I'll merge.
